pybop/_optimisation.py

```diff
@@ -73,441 +71,60 @@ def __init__(
-            self.optimiser = pybop.XNES
-        elif issubclass(self.optimiser, pints.Optimiser):
-            pass
-        else:
-            self.pints = False
-
-            if issubclass(
-                self.optimiser, (pybop.SciPyMinimize, pybop.SciPyDifferentialEvolution)
-            ):
-                self.optimiser = self.optimiser(bounds=self.bounds)
-
-            else:
-                raise ValueError("Unknown optimiser type")
-
-        if self.pints:
-            self.optimiser = self.optimiser(self.x0, self.sigma0, self.bounds)
-
-        # Check if sensitivities are required
-        self._needs_sensitivities = self.optimiser.needs_sensitivities()
-
-        # Track optimiser's f_best or f_guessed
-        self._use_f_guessed = None
-        self.set_f_guessed_tracking()
-
-        # Parallelisation
-        self._parallel = False
-        self._n_workers = 1
-        self.set_parallel()
-
-        # User callback
-        self._callback = None
-
-        # Define stopping criteria
-        # Maximum iterations
-        self._max_iterations = None
-        self.set_max_iterations()
-
-        # Minimum iterations
-        self._min_iterations = None
-        self.set_min_iterations()
-
-        # Maximum unchanged iterations
-        self._unchanged_threshold = 1  # smallest significant f change
-        self._unchanged_max_iterations = None
-        self.set_max_unchanged_iterations()
-
-        # Maximum evaluations
-        self._max_evaluations = None
+        self.optimiser = self.optimiser(
+            self.x0, self.sigma0, self.bounds, **optimiser_kwargs
+        )
 
-        # Threshold value
-        self._threshold = None
+        # Pass cost and settings to optimiser
+        self.optimiser._cost_function = self.cost
+        self.optimiser.verbose = self.verbose
 
-        # Post-run statistics
-        self._evaluations = None
-        self._iterations = None
+        # Check if minimising or maximising
+        if isinstance(self.cost, pybop.BaseLikelihood):
+            self.cost._minimising = False
+        self.optimiser._minimising = self.cost._minimising
 
-    def run(self):
+    def run(self, **optimiser_kwargs):
         """
-        Run the optimization and return the optimized parameters and final cost.
+        Run the optimisation and return the optimised parameters and final cost.
+
+        **optimiser_kwargs : optional
+            Valid SciPy option keys and their values.
 
         Returns
         -------
         x : numpy.ndarray
-            The best parameter set found by the optimization.
+            The best parameter set found by the optimisation.
         final_cost : float
             The final cost associated with the best parameters.
         """
 
-        if self.pints:
-            x, final_cost = self._run_pints()
-        elif not self.pints:
-            x, final_cost = self._run_pybop()
+        x, final_cost = self.optimiser.run(**optimiser_kwargs)
```

Comment:
I think it would be helpful to keep the API for run/optimise the same if possible. I.e. passing a cost function and initial conditions then kwargs for all other options. This really helps understand what the function requirements and interactions are.

Reply:
Outdated